### PR TITLE
Adding gcp pub sub connector

### LINKED ISSUE #11: GCP Pub/Sub Http connector
CC: https://github.com/fission/fission/issues/1663

### PR DESCRIPTION
**Pull request by RealHarshThakur**
Signed-off-by: Harsh Thakur <harshthakur9030@gmail.com>
Resolves #11 
Cannot be merged until https://github.com/kedacore/keda/issues/1273 is resolved

### CODE REVIEW

File: aws-sqs-http-connector/go.mod [hunk 1]
```diff
@@ -1,12 +1,9 @@
 module github.com/fission/keda-connectors/aws-sqs-http-connector
 
-go 1.12
+go 1.15
 
 require (
-	github.com/aws/aws-sdk-go v1.34.20
-	github.com/fission/keda-connectors/common v0.0.0-20200827103613-b77618fd08e1
+	github.com/aws/aws-sdk-go v1.35.7
```
Comment (therahulbhati) [inline aws-sqs-http-connector/go.mod:6]: I think you might have done this by mistake. Don't modify aws connector code, please feel to raise separate PR if required.

File: common/util.go [hunk 2]
```diff
@@ -25,6 +25,14 @@ type ConnectorMetadata struct {
 	SourceName    string
 }
 
+// GCPPubsubConnInfo contains the fields needed to connect to the GCP's pub sub topic queue.
+type GCPPubsubConnInfo struct {
```
Comment (therahulbhati) [inline common/util.go:29]: Common package is common code used by various message queues. This code specific to GCP please move this to gcp connector code.

File: common/util.go [hunk 3]
```diff
@@ -90,7 +98,7 @@ func HandleHTTPRequest(message string, headers http.Header, data ConnectorMetada
 		return nil, fmt.Errorf("every function invocation retry failed; final retry gave empty response. http_endpoint: %v, source: %v", data.HTTPEndpoint, data.SourceName)
 	}
 
-	if resp.StatusCode != 200 {
+	if resp.StatusCode < 200 && resp.StatusCode > 300 {
```
Comment (therahulbhati) [inline common/util.go:101]: Please raise a separate PR for common package, it would make testing less error-prone.

**Review by therahulbhati — CHANGES_REQUESTED** [review]

**Comment by harshthakur9030** [comment]
https://github.com/fission/fission/issues/1864 needs to be resolved since  kedacore/keda#1273  is resolved in it.

**Comment by RealHarshThakur** [comment]
@therahulbhati can you PTAL now?

File: gcp-pubsub-http-connector/main.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+	go func(res *pubsub.PublishResult) {
+		// The Get method blocks until a server-generated ID or
+		// an error is returned for the published message.
+		_, err := res.Get(ctx)
+		if err != nil {
+			conn.logger.Error("Failed to publish: %v", zap.Error(err))
+			return
+		}
+	}(result)
+	conn.logger.Sugar().Infof("Published message , topic name: %v\n", topicID)
```
Comment (therahulbhati) [inline gcp-pubsub-http-connector/main.go:141]: @RealHarshThakur Any particular reason to use Sugar() wrapper? Doc says it bit slower https://godoc.org/go.uber.org/zap#Logger.Sugar

Comment (therahulbhati) [reply gcp-pubsub-http-connector/main.go:141]: In other connectors, we have used key value format in the loggers.

Comment (RealHarshThakur) [reply gcp-pubsub-http-connector/main.go]: Fixed

File: gcp-pubsub-http-connector/main.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+			}
+			conn.logger.Error("Error sending the message to the endpoint %v", zap.Error(err))
+		} else {
+			body, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				if conn.connectordata.ErrorTopic != "" {
+					conn.responseOrErrorHandler(conn.connectordata.ErrorTopic, string(body), headers)
+				}
+				conn.logger.Error("Error reading body", zap.Error(err))
+				resp.Body.Close()
```
Comment (therahulbhati) [inline gcp-pubsub-http-connector/main.go:104]: Can we have body close in defer statement before we open like at line 183 like `defer resp.Body.Close()`

**Review by therahulbhati — CHANGES_REQUESTED** [review]

File: gcp-pubsub-http-connector/main.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+	conn.logger.Info("Published message , topic name: %v\n", zap.String("Topic", topicID))
+
+}
+
+// convHeadersToAttr converts the headers to attributes which can be published by pubsub
+func convHeadersToAttr(headers http.Header) map[string]string {
+	var attr = make(map[string]string)
+	for k, v := range headers {
+		for _, d := range v {
+			attr[k] = d
```
[the remark this answers is not in the document]
Comment (RealHarshThakur) [reply gcp-pubsub-http-connector/main.go:150]: Users can set only one value for each key. The only problematic keys are the keys we set in headers. As for those, if user sets a value to it, we over ride that with what the user has given as a result of this function.

File: gcp-pubsub-http-connector/main.go [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+		// Set the attributes as message header came from PubSub record
+		for k, v := range msg.Attributes {
+			headers.Add(k, v)
+		}
+
+		// Push the message to the endpoint
+		resp, err := common.HandleHTTPRequest(string(msg.Data), headers, conn.connectordata, conn.logger)
+		if err != nil {
+			if conn.connectordata.ErrorTopic != "" {
+				conn.responseOrErrorHandler(conn.connectordata.ErrorTopic, "Error", headers)
```
Comment (therahulbhati) [inline gcp-pubsub-http-connector/main.go:94]:
```suggestion
				conn.responseOrErrorHandler(conn.connectordata.ErrorTopic, err.Error(), headers)
```

File: gcp-pubsub-http-connector/main.go [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+		conn.logger.Error("pubsub.NewClient: %v", zap.Error(err))
+	}
+
+	t := client.Topic(topicID)
+	result := t.Publish(ctx, &pubsub.Message{
+		Data:       []byte(response),
+		Attributes: convHeadersToAttr(headers),
+	})
+
+	go func(res *pubsub.PublishResult) {
```
[the remark this answers is not in the document]
Comment (therahulbhati) [reply gcp-pubsub-http-connector/main.go:133]: Can we use waitGroup

**Review by therahulbhati — CHANGES_REQUESTED** [review]

**Review by therahulbhati — APPROVED** [review]
LGTM

**Pull request merged**